utils/serde/src/main/java/io/apicurio/registry/utils/serde/JsonSchemaKafkaDeserializer.java

```diff
@@ -0,0 +1,217 @@
+            @Override
+            protected SchemaValidator toSchema(Response response) {
+                String schema = response.readEntity(String.class);
+                return api.loadSchema(new StringSchemaSource(schema));
+            }
+        };
+    }
+    
+    /**
+     * @see org.apache.kafka.common.serialization.Deserializer#deserialize(java.lang.String, byte[])
+     */
+    @Override
+    public T deserialize(String topic, byte[] data) {
+        throw new UnsupportedOperationException();
+    }
+    
+    /**
+     * @see org.apache.kafka.common.serialization.Deserializer#deserialize(java.lang.String, org.apache.kafka.common.header.Headers, byte[])
+     */
+    @Override
+    public T deserialize(String topic, Headers headers, byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        
+        try {
+            JsonParser parser = mapper.getFactory().createParser(data);
+            if (validationEnabled) {
+                Long globalId = getGlobalId(headers);
+                
+                // If no globalId is provided, check the alternative - which is to check for artifactId and 
+                // (optionally) version.  If these are found, then convert that info to globalId.
+                if (globalId == null) {
+                    String artifactId = getArtifactId(headers);
+                    Integer version = getVersion(headers);
+                    if (version == null) {
+                        version = getLatestVersion(artifactId);
+                    }
+                    globalId = toGlobalId(artifactId, version);
+                }
+                
+                SchemaValidator schema = schemaCache.getSchema(globalId);
+                parser = api.decorateJsonParser(schema, parser);
+            }
+            
+            Class<T> messageType = getMessageType(headers);
+
+            return mapper.readValue(parser, messageType);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Gets the global id from the headers.  Returns null if not found.
+     * @param headers
+     */
+    protected Long getGlobalId(Headers headers) {
+        Header header = headers.lastHeader(JsonSchemaSerDeConstants.HEADER_GLOBAL_ID);
+        if (header == null) {
+            return null;
+        }
+        ByteBuffer buffer = ByteBuffer.allocate(8);
+        buffer.put(header.value());
+        buffer.position(0);
+        return buffer.getLong();
+    }
+
+    /**
+     * Gets the artifact id from the headers.  Throws if not found.
+     * @param headers
+     */
+    protected String getArtifactId(Headers headers) {
+        Header header = headers.lastHeader(JsonSchemaSerDeConstants.HEADER_ARTIFACT_ID);
+        if (header == null) {
+            throw new RuntimeException("ArtifactId not found in headers.");
+        }
+        return new String(header.value(), StandardCharsets.UTF_8);
```

Comment:
Use IoUtils to handle byte[] --> String ... 
(a single place to handle such things)

Reply:
I think using the String constructor is superior in this case.  The `toString(byte[], String)` method declares an `IOException` (even though it can't throw one) and then just calls the same String c'tor that I'm using here.  It's even slightly worse though because it parses the Charset from string rather than just using one.

So -1 for me on this change. :)

Reply:
Ah, it's our IoUtil:

    public static byte[] toBytes(String string) {
        return (string == null ? null : string.getBytes(StandardCharsets.UTF_8));
    }

It's just that we have the selection of charset in a single place / class.

Reply:
Oh I see.  Didn't notice that class.  I've made that change in all relevant places (I think).